File: algorithms/tree/design/binary_search_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, value) -> None:
        if self.root is None:
            self.root = TreeNode(value)
            return

        curr = self.root
        while curr:
            if value < curr.val:
                if curr.left is None:
                    curr.left = TreeNode(value)
                    break
                curr = curr.left
            elif value > curr.val:
                if curr.right is None:
                    curr.right = TreeNode(value)
                    break
                curr = curr.right
            else:
                break

    def search(self, value) -> TreeNode | None:
        curr = self.root
        while curr:
            if curr.val == value:
                return curr
            elif value < curr.val:
                curr = curr.left
            else:
                curr = curr.right

        return None

    def delete(self, value):
        self.root = self._delete_recursive(self.root, value)

    def _min_value_node(self, node):
        curr = node

        while curr.left:
            curr = curr.left

        return curr

    def _delete_recursive(self, node, value):
        if not node:
            return None

        if value < node.val:
            node.left = self._delete_recursive(node.left, value)
        elif value > node.val:
            node.right = self._delete_recursive(node.right, value)
        else:
            if not node.left:
                return node.right
            elif not node.right:
                return node.left

            successor = self._min_value_node(node.right)
            node.val = successor.val

            node.right = self._delete_recursive(node.right, successor.val)

        return node

File: cpp_algorithms/tree/design/binary_search_tree.cpp
#include <bits/stdc++.h>
using namespace std;

// Binary Search Tree:
//
// Time Complexity:
// insert(): Average O(log n), Worst O(n)
// search(): Average O(log n), Worst O(n)
// delete(): Average O(log n), Worst O(n)
//
// Space Complexity:
// Memory: O(n)
// insert(): O(1)
// search(): O(1)
// delete(): recursive O(h)

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int value) {
        val = value;
        left = nullptr;
        right = nullptr;
    }
};

class BinarySearchTree {
private:
    TreeNode* root;

    TreeNode* minValueNode(TreeNode* node) {
        TreeNode* curr = node;

        while (curr->left != nullptr) {
            curr = curr->left;
        }

        return curr;
    }

    TreeNode* deleteRecursive(TreeNode* node, int value) {
        if (node == nullptr) {
            return nullptr;
        }

        if (value < node->val) {
            node->left = deleteRecursive(node->left, value);
        }
        else if (value > node->val) {
            node->right = deleteRecursive(node->right, value);
        }
        else {
            if (node->left == nullptr) {
                TreeNode* rightChild = node->right;
                delete node;
                return rightChild;
            }

            if (node->right == nullptr) {
                TreeNode* leftChild = node->left;
                delete node;
                return leftChild;
            }

            TreeNode* successor = minValueNode(node->right);

            node->val = successor->val;

            node->right = deleteRecursive(
                node->right,
                successor->val
            );
        }

        return node;
    }

public:
    BinarySearchTree() {
        root = nullptr;
    }

    void insert(int value) {
        if (root == nullptr) {
            root = new TreeNode(value);
            return;
        }

        TreeNode* curr = root;

        while (curr != nullptr) {
            if (value < curr->val) {
                if (curr->left == nullptr) {
                    curr->left = new TreeNode(value);
                    break;
                }

                curr = curr->left;
            }
            else if (value > curr->val) {
                if (curr->right == nullptr) {
                    curr->right = new TreeNode(value);
                    break;
                }

                curr = curr->right;
            }
            else {
                // Такое значение уже есть
                break;
            }
        }
    }

    TreeNode* search(int value) {
        TreeNode* curr = root;

        while (curr != nullptr) {
            if (curr->val == value) {
                return curr;
            }
            else if (value < curr->val) {
                curr = curr->left;
            }
            else {
                curr = curr->right;
            }
        }

        return nullptr;
    }

    void remove(int value) {
        root = deleteRecursive(root, value);
    }

    TreeNode* getRoot() {
        return root;
    }
};
